File: SMSnpcasimu/SMSnpcasimu.m
clear all
close all
clc

RandStream.setGlobalStream(RandStream('mt19937ar','seed',1));

u         = unifrnd(0,1,50,2);
a         = [1, 0, 1, 1
             0, 1, 1, -1];
  
x         = u * a;                     % generate the random data set

x         = zscore(x);                 % normalizing data
n         = size(x,1);
[v e]     = eigs(cov(x));              % spectral analysis
e         = diag(e);
tmp       = [e,v'];
e         = tmp(:,1);
v         = tmp(:,2:5)';
v(:,[1,4])= -v(:,[1,4]);               % sign correction for the book

disp('eigenvalues and proportions of explained variance');
disp([e , (e/sum(e)) , (cumsum(e)/sum(e))]');
disp('2 eigenvectors')
disp(v(:,1:2)');

y      = x * v;    % principal components
x12    = y(:,1:2);                  % first two PCs

figure(1)
subplot(1,2,1)                      % scatterplot of first two PCs
plot(x12(:,1),x12(:,2),'ob')
title('first  vs. second PC','FontSize',16,'FontWeight','Bold')
xlabel('PC 1','FontSize',16,'FontWeight','Bold')
ylabel('PC 2','FontSize',16,'FontWeight','Bold')
set(gca,'LineWidth',1.6,'FontSize',16,'FontWeight','Bold')
box on
subplot(1,2,2)                      % screeplot of eigenvalues
bar(e)
ylim([0,2.5])
xlim([0,5])
title('eigenvalues of S','FontSize',16,'FontWeight','Bold')
ylabel('\lambda','FontSize',16,'FontWeight','Bold')
box on
set(gca,'LineWidth',1.6,'FontSize',16,'FontWeight','Bold')


% To save plots uncomment following lines
% print -painters -dpdf -r600 SMSnpcasimu01.pdf
% print -painters -dpng -r600 SMSnpcasimu01.png


r       = corr([y,x]);
r       = r(5:8,1:2);                % correlations between variables and pc's
ucircle = [cos((0:360)/180*3.14159)',sin((0:360)/180*3.14159)'];
figure(2)
plot(ucircle(:,1),ucircle(:,2), '-w')
title('simulated data set','FontSize',16,'FontWeight','Bold')
xlabel('first PC','FontSize',16,'FontWeight','Bold')
ylabel('second PC','FontSize',16,'FontWeight','Bold')
hold on
xlim([-1.2 1.2])
ylim([-1.2 1.2])
line([-1.2 1.2],[0 0],'Color','k')
line([0 0],[1.2 -1.2],'Color','k')
line(ucircle(:,1),ucircle(:,2),'LineStyle',':','LineWidth',1.6);
plot(r(:,1),r(:,2),'ok');
text(r(:,1),r(:,2),  cellstr(strvcat('X1','X2','X3','X4')),'FontSize',16,'FontWeight','Bold')
box on
set(gca,'LineWidth',1.6,'FontSize',16,'FontWeight','Bold')

% To save plots uncomment following lines
% print -painters -dpdf -r600 SMSnpcasimu02.pdf
% print -painters -dpng -r600 SMSnpcasimu02.png

File: SMSnpcasimu/SMSnpcasimu.R
# clear cache and close windows
rm(list=ls(all=TRUE))
graphics.off()

# fix pseudo random numbers
set.seed(1)

u   = matrix(runif(100), ncol = 2);
a   = cbind(c(1,0), c(0,1), c(1,1), c(1, -1));
x   = u %*% a;                     # generate the random data set
x   = scale(x);
n   = nrow(x);
tmp = eigen(cov(x), symmetric = TRUE);
e   = tmp$values;
v   = tmp$vectors;                 # spectral analysis
tmp = cbind(e,t(v))[order(-e),];
e   = tmp[,1];
v   = t(tmp[,2:5]);

print("eigenvalues and proportions of explained variance");
print(t(cbind(e,(e/sum(e)),(cumsum(e)/sum(e)))));
print("2 eigenvectors")
print(t(v[,1:2]));

y   = x %*% v;                     # principal components
x12 = y[,1:2];

op = par(mfrow = c(1,2));
plot(x12, main = "first  vs. second PC", xlab = "pc1", ylab = "pc2");
plot(cbind(c(1:4),e), main = "eigenvalues of S", xlab = "", ylab = expression(lambda));
# screeplot of eigenvalues
par(mfrow = c(1,1),ask = dev.interactive(orNone = TRUE))
r = cor(cbind(y,x));
r = r[5:8,1:2];                  # correlations between variables and pc's
ucircle = cbind(cos((0:360)/180*3.14159),sin((0:360)/180*3.14159));
xline = cbind(c(-1.1,1.1),c(0,0));
yline = cbind(c(0,0),c(-1.1,1.1));
dev.new()
plot(ucircle, type = "n", main = "simulated data set", xlab = "first PC", ylab = "second PC");
lines(ucircle, lty = "dashed");
lines(xline, lwd = 2);
lines(yline, lwd = 2);
points(r);
text(r, c("X1","X2","X3","X4"), adj = 1.2);
par(op);
